### ISSUE TITLE
License compliance tooling for core, federated extensions

### ISSUE THREAD

**Issue by bollwyvl** [jupyterlab/jupyterlab#9518]
### Problem

Being able to package extensions of arbitrary complexity in pip/conda is great for users! Having to manually collect licenses for all of the included, minified asset, however, will not be fun for developers and re-packagers unless highly automated.

### Proposed Solution

- Add [license-webpack-plugin](https://www.npmjs.com/package/license-webpack-plugin) or similar to `@jupyterlab/builder` to dump all of the licenses into a well-known file location in the resulting `static` folder. 
- Expose all of these things 
  - in the CLI, e.g. `jupyterlab labextension list --show-licenses` and/or
  - all the way up to the UI, e.g. _Help &raquo; About &raquo; Open Source Licenses_, fetch all the licenses from all known extensions from a `jupyterlab_server` endpoint.

### Additional context

Downstream on conda-forge, license compliance is kind of a big deal, and demonstration of compliance helps litigation-risk-averse organizations adopt open source. Further, core probably shouldn't ship, say, AGPL JS (if such things exist), etc. so having some rules in-place for core extensions might help evaluate possible new dependencies.

**Comment by jasongrout** [comment]
Dup of #9140

**Comment by jasongrout** [comment]
You explored things a bit further in this issue, and have an accompanying PR, so I'll close the other issue in favor of this one. Thanks!

**Comment by krassowski** [comment]
@bollwyvl can we close this one, or is there anything else to do? I see the PR was merged and released in 3.1.

**Comment by bollwyvl** [comment]
Yeah, we landed all the plumbing, the 3.x line now ships with `third-party-licenses.json`, and all extensions built with `@jupyterlab/builder: ^3.1.0` will include the third party license information. So hooray!

What we did _not_ add:

> rules in-place for core extensions might help evaluate possible new dependencies.

This would probably take:
- determine a list of licenses which core _may not ship_
  - IANAL, but 
    - imagine AGPL is among these
    - as are some of the new-fangled, untried-in-court "ethical source" licenses... good luck enforcing "don't be evil,"
- add a check in `ensure-repo` or somewhere to validate
  - nothing from the unshippables is on the list
  - nothing _without_ a declared license is shipped, as it could secretly _be_ an unshippable

If we _do_ this due diligence, then we'd probably have to _act_ on any findings :rofl:

**Comment by bollwyvl** [comment]
So I will admit to having looked :rofl: This is the reduced list (simplifying some `OR` LicenseRefs) of JS licenses we currently ship, and there are no `null`s:

```py
SHIPPED_LICENSES = [
    r"0BSD",
    r"Apache-2\.0",
    r"BSD-(2|3)-Clause",
    r"CC0-1.0"
    r"ISC",
    r"MIT",
]
```

So it may be easier to fiat _these are the licenses we ship_, start testing them, and address novel dependencies on a case-by-case basis.

**Comment by bollwyvl** [comment]
I guess I don't know where the test would actually land:

- jupyter_releaser?
- python unit tests?
- random `script`?

Similarly, the list of licenses seems... important. It should end up in _exactly one place_, but that could be:

- `package.json`
- `pyproject.toml`
- some random `script`
- the docs

Thoughts?

**Comment by krassowski** [comment]
I would put that test together with repo integrity check. I think that running it on release may be to late, it would be better to be able to catch issues with new dependencies at a PR stage.

As for the place documenting the third-party licenses, the docs seem natural to me.
